app/users/authenticator.go

```diff
@@ -0,0 +1,62 @@
+package users
+
+import (
+	"errors"
+	"net/http"
+)
+
+type Authenticator struct {
+	retriever Retriever
+}
+
+type AuthenticatedRequest struct {
+	*http.Request
+	AccountID string
+	AuthError error
+}
+
+type AuthenticatedFunc func(http.ResponseWriter, *AuthenticatedRequest)
+
+// NewAuthenticator provides HTTP handler wrapping methods
+// and should be initialized with an object that allows user retrieval.
+func NewAuthenticator(retriever Retriever) *Authenticator {
+	return &Authenticator{retriever}
+}
+
+// GetAccountID retrieves user token from HTTP headers and subsequently
+// an SDK account ID from Retriever.
+func (a *Authenticator) GetAccountID(r *http.Request) (string, error) {
+	if token, ok := r.Header[TokenHeader]; ok {
+		u, err := a.retriever.Retrieve(token[0])
+		if err != nil {
+			return "", err
+		}
+		if u == nil {
+			return "", errors.New("unable to retrieve user")
+		}
+		return u.SDKAccountID, nil
+	}
+	return "", nil
+}
+
+// Wrap result can be supplied to all functions that accept http.HandleFunc,
+// supplied function will be wrapped and called with AuthenticatedRequest instead of http.Request.
+func (a *Authenticator) Wrap(wrapped AuthenticatedFunc) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		AccountID, err := a.GetAccountID(r)
+		ar := &AuthenticatedRequest{Request: r}
+		if err != nil {
+			ar.AuthError = err
+		} else {
+			ar.AccountID = AccountID
```

Comment:
can the accountID ever be nil or invalid? perhaps you need to check that too?

Reply:
nah, it's either retrieved from the DB or it isn't, in that case the `err` will be set.